### PR TITLE
Refactoring of tools

### PR DESCRIPTION
**Pull request by vgkholla**
1. Refactored AdminTool (ServerAdminTool) to use NetworkClient and added support
for triggering admin operations.
2. Refactored and cleaned up BlobValidator making it use the ServerAdminTool (code reuse).

### CODE REVIEW

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Tool to support admin related operations on Ambry server
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: can we add java docs here to list the operations that this tool supports.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 2]
```diff
@@ -16,654 +16,445 @@
...
+     */
+    @Config("get.option")
+    @Default("None")
+    final GetOption getOption;
+
+    /**
+     * Replicas to contact per second (for throttling_
+     */
+    @Config("replicas.to.contact.per.sec")
+    @Default("None")
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: can we set default to 1

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 3]
```diff
@@ -16,654 +16,445 @@
...
-      if (connectionPool != null) {
-        connectionPool.shutdown();
+        blobIdsFilePath = null;
+        blobIds = verifiableProperties.getString("blob.ids").split(",");
       }
+      hostname = verifiableProperties.getString("hostname", "localhost");
+      port = verifiableProperties.getIntInRange("port", 6667, 1, 65535);
+      datacenter = verifiableProperties.getString("datacenter", "");
+      getOption = GetOption.valueOf(verifiableProperties.getString("get.option", "None"));
+      replicasToContactPerSec =
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: have you tried running the tool with all different inputs (diff operations, diff get options, blobIdfilePath and so on). If you don't have time, let me know. I will try to run it to ensure everything works fine.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    if (!file.exists() && !file.createNewFile()) {
+      throw new IllegalStateException("Could not create " + file);
+    }
+    FileOutputStream outputFileStream = new FileOutputStream(config.dataOutputFilePath);
+    DataNodeId dataNodeId = clusterMap.getDataNodeId(config.hostname, config.port);
+    if (dataNodeId == null) {
+      throw new IllegalArgumentException(
+          "Could not find a data node corresponding to " + config.hostname + ":" + config.port);
+    }
+    switch (config.typeOfOperation) {
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: again, same question as other class. Did you get to try out all diff operations. Re-iterating again, bcoz, you have revamped the whole tool. So, just wanted to ensure.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 5]
```diff
@@ -16,654 +16,445 @@
...
+    /**
+     * The get option to use to do the get operation (if applicable)
+     * Applicable for: GetBlobProperties,GetUserMetadata,GetBlob
+     */
+    @Config("get.option")
+    @Default("None")
+    final GetOption getOption;
+
+    /**
+     * Replicas to contact per second (for throttling_
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: was a typo - fixed

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java (hunk 2 above)
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: It is one in the actual code - looks like I had not updated the doc. Fixed now.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java (hunk 3 above)
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: I tried all three operations, tried blob list and with file path, tried `GetOptions.None` and `GetOptions.Include_All` (not a full combination but at least one test with each config). That covers most of the cases I think.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java (hunk 1 above)
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: Added.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java (hunk 4 above)
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: Tried all four operations, ensured that blob data and user metadata written to output file matches what was put and tried `GetOptions.None` and `GetOptions.Include_All`.

**Comment by vgkholla** [comment]
@nsivabalan Addressed comments.

**Comment by vgkholla** [comment]
@nsivabalan Gentle reminder to review

**Comment by nsivabalan** [comment]
Looks good. Merging.

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java [hunk 6]
```diff
@@ -245,11 +245,18 @@ public static void main(String args[]) throws Exception {
...
         break;
       case TriggerCompaction:
-        ServerErrorCode errorCode = serverAdminTool.triggerCompaction(dataNodeId, config.partitionId, clusterMap);
-        if (errorCode == ServerErrorCode.No_Error) {
-          LOGGER.info("Compaction has been triggered for {} on {}", config.partitionId, dataNodeId);
+        if (config.partitionIds.length > 0 && !config.partitionIds[0].isEmpty()) {
+          for (String partitionId : config.partitionIds) {
+            ServerErrorCode errorCode = serverAdminTool.triggerCompaction(dataNodeId, partitionId, clusterMap);
+            if (errorCode == ServerErrorCode.No_Error) {
+              LOGGER.info("Compaction has been triggered for {} on {}", partitionId, dataNodeId);
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: Do you think having an option whether to proceed on failure or throw an exception would help.

Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/ServerAdminTool.java]: I think that they are independent events and should go on - errors can be dealt with later

**Pull request merged**